### PR TITLE
Test branch

### PR DESCRIPTION
**Pull request by KonstantinZhukovskij**

### CODE REVIEW

File: client/src/components/SelectLanguage/index.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+const SelectLanguage = ({ onChangeLanguage }) => (
+    <div className='btn-group dropup'>
+        <button type='button' className='btn dropdown-toggle languageButton' data-toggle='dropdown'
+                aria-haspopup='true' aria-expanded='false'>
+            {TranslationService.getTranslation('languageButton')}
+        </button>
+        <div className='dropdown-menu'>
+            <button className='dropdown-item chooseRussian'
+                    onClick={() => onChangeLanguage('ru_RU')}>{TranslationService.getTranslation('chooseRussian')}
```
Comment (nikolyachikm) [inline client/src/components/SelectLanguage/index.js]: `translationService.changeLanguage(...)`
```
translationService:

import localisationRu from 'ru_RU.json';
import localisationEn from 'en_EN.json';

changeLanguage(language): {
localStorage.setItem ....

switch language {
case 'ru_RU':  // <-move to constants
this.data = localisationRu;
}
default:
this.data = localisationEn;
}
```

File: client/src/components/YandexMap/index.js [hunk 2]
```diff
@@ -79,39 +82,50 @@ class YandexMap extends React.Component {
...
-                data: { content: localStorage.key(index).split(' ')[1] },
+                data: { content: city[index].city },
                 options: { selectOnClick: false },
-                coordinates: [Number(localStorage.getItem(localStorage.key(index)).split(' ')[1]),
-                    Number(localStorage.getItem(localStorage.key(index)).split(' ')[0])]
+                coordinates: [city[index].latitude, city[index].longitude]
             })
         }
-        return lastCities
+        return this.deleteSameObjectsInArray(lastCities)
```
Comment (nikolyachikm) [inline client/src/components/YandexMap/index.js]: you should not add duplicates in array instead of filter them during `get`

Comment (nikolyachikm) [reply client/src/components/YandexMap/index.js]: also it's possibility to you to make acquaintance with fixers: write service "fixer" that will load once for user (even if he will reload page multiple times - fixer should be runned once on first load of page) and will remove duplicates in cities. name it DuplicateCitiesFixer. if you need any details - let me know

File: client/src/dictionary/eng.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "cityTitle": "Find out the weather in your city",
```
Comment (nikolyachikm) [inline client/src/dictionary/eng.json]: {
seachForm: {
"FindOutTheWeatherInYourCity": "Find out the weather in your city"
...
}
...
}

File: client/src/dictionary/eng.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  "checkWeatherButton": "Check weather",
+  "form-control": "Enter your city here",
+  "listBox": "Last cities",
+  "futureWeatherButton": "Show future weather",
+  "chooseEnglish": "English",
+  "chooseRussian": "Russian",
+  "languageButton": "Language",
+  "balloonContentHeader": "Weather today",
+  "balloonContentFooter": "Click <strong>Show future weather</strong> to check the weather for the next days",
```
Comment (nikolyachikm) [inline client/src/dictionary/eng.json]: hm... is there any other way to not keep html in text ?

File: client/src/index.js [hunk 5]
```diff
@@ -1,6 +1,55 @@
-import React from 'react';
-import ReactDOM from 'react-dom';
 import App from 'App';
 import 'App.scss';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+const urlBase64ToUint8Array = (base64String) => {
```
Comment (nikolyachikm) [inline client/src/index.js]: move to service

File: client/src/styles/selectLanguage.css [hunk 6]
```diff
@@ -0,0 +1,7 @@
+.dropup {
```
Comment (nikolyachikm) [inline client/src/styles/selectLanguage.css]: project should not contain .css and .scss together

File: server/routers/index.js [hunk 7]
```diff
@@ -1,32 +1,76 @@
...
-getPlaceData = (request, response) => {
-    const city = encodeURIComponent(request.query.city);
-    return axios.get(`https://geocode-maps.yandex.ru/1.x/?format=json&lang=ru_RU&apikey=990aeae6-c5b9-4279-af0c-74ac3beb30d2&geocode=${city}`)
-        .then((result) => {
-            return response.json(result.data.response)
-        })
-        .catch((error) => {
-            console.log("error: ", error)
-        })
+SaveDataInMemoryCache = (key, object) => {
```
Comment (nikolyachikm) [inline server/routers/index.js]: saveDataInMemoryCache

Comment (nikolyachikm) [reply server/routers/index.js]: and move to another service.

all components, services, and other  should follow SOLID principles

File: client/src/components/MainPage/index.js [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
+import moment from 'moment';
+import React, { Component } from 'react';
 import { getPlaceData, getWeatherData } from 'services/ajaxSerivce';
+import toastr from 'services/toastrSerivce';
 
 export default class MainPage extends Component {
     state = {
         city: '',
-        dateTime: moment().format('MMMM Do YYYY, h:mm a'),
+        dateTime: moment().format('DD.MM.YYYY HH:mm'),
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: I use moment library in more places of my application, so I used it and here

File: client/src/components/MainPage/index.js [hunk 9]
```diff
@@ -49,15 +49,23 @@ export default class MainPage extends Component {
 
         this.setState({ isLoading: true });
 
-        getPlaceData(this.state.city)
+        getPlaceData(this.state.city, this.props.language)
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/dictionary/eng.json (hunk 3 above)
Comment (KonstantinZhukovskij) [reply client/src/dictionary/eng.json]: Done

File: client/src/App.js [hunk 10]
```diff
@@ -1,16 +1,30 @@
...
 
-class App extends Component {
+export default class App extends Component {
+    state = {
+        language: ((localStorage.getItem('language') === null)
+            ? localStorage.setItem('language', 'en_RU')
+            : localStorage.getItem('language'))
+    };
+
+    onChangeLanguage = (language) => {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/App.js]: Done

File: client/src/App.js [hunk 11]
```diff
@@ -1,16 +1,30 @@
+import CloudBackground from 'components/CloudBackground';
+import MainPage from 'components/MainPage';
+import SelectLanguage from 'components/SelectLanguage';
 import React, { Component } from 'react';
-import MainPage from "components/MainPage";
-import CloudBackground from "components/CloudBackground";
 
-class App extends Component {
+export default class App extends Component {
+    state = {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/App.js]: Done

File: client/src/index.js (hunk 5 above)
Comment (KonstantinZhukovskij) [reply client/src/index.js]: Done

File: client/src/components/SearchForm/index.js [hunk 12]
```diff
@@ -1,5 +1,6 @@
-import React, { Component } from 'react';
 import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import TranslationService from 'services/translationService'
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/SearchForm/index.js]: Done

File: client/src/components/SelectLanguage/index.js (hunk 1 above)
Comment (KonstantinZhukovskij) [reply client/src/components/SelectLanguage/index.js]: Done

File: client/src/dictionary/eng.json (hunk 4 above)
Comment (KonstantinZhukovskij) [reply client/src/dictionary/eng.json]: I don't know, but it works

File: client/src/components/MainPage/index.js [hunk 13]
```diff
@@ -49,15 +49,23 @@ export default class MainPage extends Component {
...
+                const locationCoordinates = featureMember[0].GeoObject.Point.pos;
                 const latitude = locationCoordinates.split(' ')[1];
                 const longitude = locationCoordinates.split(' ')[0];
+                const newCity = {
+                    city: city,
+                    latitude: latitude,
+                    longitude: longitude
+                };
+
+                const savedCities = JSON.parse(localStorage.getItem('cities'));
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/SearchForm/index.js [hunk 14]
```diff
@@ -11,15 +12,16 @@ export default class SearchForm extends Component {
...
-                        <input className='form-control' type="text" placeholder='Enter your city here'
+                        <input className='form-control' type='text'
+                               placeholder={TranslationService.getTranslation('form-control')}
                                onChange={this.props.onCityChange}
                                onKeyPress={(event) => (event.charCode === 13) && this.props.loadLocationAndWeatherData()}
                                value={this.props.city}/>
                         <div className='input-group-append'>
-                            <button className='btn btn-outline-secondary'
-                                    onClick={this.props.loadLocationAndWeatherData}>Check weather
+                            <button className='btn btn-outline-secondary checkWeatherButton'
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/SearchForm/index.js]: -

File: client/src/components/YandexMap/index.js (hunk 2 above)
Comment (KonstantinZhukovskij) [reply client/src/components/YandexMap/index.js]: Done

File: server/routers/index.js (hunk 7 above)
Comment (KonstantinZhukovskij) [reply server/routers/index.js]: Done

File: client/src/styles/selectLanguage.css (hunk 6 above)
Comment (KonstantinZhukovskij) [reply client/src/styles/selectLanguage.css]: Done

File: client/src/components/MainPage/index.js [hunk 15]
```diff
@@ -49,15 +51,20 @@ export default class MainPage extends Component {
...
+                const locationCoordinates = featureMember[0].GeoObject.Point.pos;
                 const latitude = locationCoordinates.split(' ')[1];
                 const longitude = locationCoordinates.split(' ')[0];
+                const newCity = {
+                    city: city,
+                    latitude: latitude,
+                    longitude: longitude
+                };
+
+                localStorageService.moveObjectToEndOfArray(newCity);
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: localStorageService.addCity or localStorageService.updateCity

File: client/src/components/MainPage/index.js [hunk 16]
```diff
@@ -88,12 +95,14 @@ export default class MainPage extends Component {
     };
 
     showMapIfUserWasEarlier = () => {
-        if (localStorage.length >= 1) {
+        if (localStorageService.getDataToLocalStorage('cities') !== null) {
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: localStorageService.getCities

File: client/src/components/MainPage/index.js [hunk 17]
```diff
@@ -88,12 +95,14 @@ export default class MainPage extends Component {
...
             this.setState({ isLoading: true });
 
-            const city = localStorage.key(localStorage.length - 1).split(' ')[1];
-            const latitude = localStorage.getItem(localStorage.key(localStorage.length - 1)).split(' ')[1];
-            const longitude = localStorage.getItem(localStorage.key(localStorage.length - 1)).split(' ')[0];
+            const arrayCities = localStorageService.saveDataToLocalStorage('cities');
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: localStorageService.updateCities. also you don't pass anything here - why? it's "save" method -we should pass something here

File: client/src/components/SearchForm/index.js [hunk 18]
```diff
@@ -11,15 +12,16 @@ export default class SearchForm extends Component {
         return (
             <div className='row'>
                 <div className='col-12'>
-                    <h4 className='cityTitle'>Find out the weather in your city</h4>
+                    <h4 className='cityTitle'>{translationService.getTranslation('findOutTheWeatherInYourCity')}</h4>
```
Comment (nikolyachikm) [inline client/src/components/SearchForm/index.js]: I asked you to remove calls of functions to get text

File: client/src/components/MainPage/index.js (hunk 15 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 16 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 17 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: -

File: client/src/components/SearchForm/index.js (hunk 18 above)
Comment (KonstantinZhukovskij) [reply client/src/components/SearchForm/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 19]
```diff
@@ -88,14 +95,16 @@ export default class MainPage extends Component {
...
 
     showMapIfUserWasEarlier = () => {
-        if (localStorage.length >= 1) {
+        if (localStorageService.getCities('cities') !== null) {
             this.setState({ isLoading: true });
 
-            const city = localStorage.key(localStorage.length - 1).split(' ')[1];
-            const latitude = localStorage.getItem(localStorage.key(localStorage.length - 1)).split(' ')[1];
-            const longitude = localStorage.getItem(localStorage.key(localStorage.length - 1)).split(' ')[0];
+            const arrayCities = localStorageService.getCities('cities');
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: why we need 'cities' here as param ?

File: client/src/components/YandexMap/index.js [hunk 20]
```diff
@@ -81,37 +83,40 @@ class YandexMap extends React.Component {
...
-            lastCities.push({
-                data: { content: localStorage.key(index).split(' ')[1] },
-                options: { selectOnClick: false },
-                coordinates: [Number(localStorage.getItem(localStorage.key(index)).split(' ')[1]),
-                    Number(localStorage.getItem(localStorage.key(index)).split(' ')[0])]
-            })
+        const arrayOfAllCities = JSON.parse(localStorage.getItem('cities'));
+        if (arrayOfAllCities) {
+            for (let index = 0; index < arrayOfAllCities.length; index++) {
+                let city = JSON.parse(localStorage.getItem('cities'));
```
Comment (nikolyachikm) [inline client/src/components/YandexMap/index.js]: localStorageService

File: client/src/dictionary/eng.json [hunk 21]
```diff
@@ -0,0 +1,15 @@
+{
+  "findOutTheWeatherInYourCity": "Find out the weather in your city",
```
Comment (nikolyachikm) [inline client/src/dictionary/eng.json]: better to split text by components :

serachForm: {
...
},
maps: {
...
},
etc.

File: client/src/services/localStorageService.js [hunk 22]
```diff
@@ -0,0 +1,33 @@
+class LocalStorageService {
+    getData = (key) => {
+        return localStorage.getItem(key)
+    };
+
+    setData = (key, data) => {
+        return localStorage.setItem(key, data)
+    };
+
+    getCities = (key) => {
```
Comment (nikolyachikm) [inline client/src/services/localStorageService.js]: no need in key here

File: client/src/services/localStorageService.js [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+    updateCity = (city) => {
+        const savedCities = JSON.parse(localStorage.getItem('cities'));
+
+        if (savedCities) {
+            savedCities.forEach((element) => {
+                if (city.latitude + city.longitude === element.latitude + element.longitude) {
+                    savedCities.push(savedCities.splice(savedCities.indexOf(element), 1)[0]);
+
+                    localStorage.setItem('cities', JSON.stringify([...savedCities]))
```
Comment (nikolyachikm) [inline client/src/services/localStorageService.js]: 'cities' - move to constants

File: client/src/services/translationService.js [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+    constructor() {
+        this.language = ((localStorageService.getData('language') === null)
+            ? localStorageService.setData('language', 'en_RU')
+            : localStorageService.getData('language'));
+
+        this.updateData();
+    };
+
+    updateData = () => {
+        if (this.language === 'en_RU') {
```
Comment (nikolyachikm) [inline client/src/services/translationService.js]: 'en_RU' move to constants

File: client/src/services/translationService.js [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+
+        this.dictionary = russianDictionary;
+    };
+
+    getLanguage = () => {
+        return this.language
+    };
+
+    onChangeLanguage = (language) => {
+        localStorageService.setData('language', language);
```
Comment (nikolyachikm) [inline client/src/services/translationService.js]: 'language' to constants

File: server/routers/index.js [hunk 26]
```diff
@@ -1,32 +1,59 @@
-const express = require('express');
 const axios = require('axios');
+const express = require('express');
+const apiKeys = require('../configuration/secretApiKeys');
+const memoryCacheService = require('../services/memoryCacheService');
+const pushService = require('../services/pushService');
 
 const router = express.Router();
 
+getSubscriptionDataFromClient = (request, response) => {
```
Comment (nikolyachikm) [inline server/routers/index.js]: saveSubscriptionDataFromClient

Comment (KonstantinZhukovskij) [reply server/routers/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 19 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/YandexMap/index.js (hunk 20 above)
Comment (KonstantinZhukovskij) [reply client/src/components/YandexMap/index.js]: Done

File: client/src/services/localStorageService.js (hunk 22 above)
Comment (KonstantinZhukovskij) [reply client/src/services/localStorageService.js]: Done

File: client/src/services/localStorageService.js (hunk 23 above)
Comment (KonstantinZhukovskij) [reply client/src/services/localStorageService.js]: Done

File: client/src/services/translationService.js (hunk 24 above)
Comment (KonstantinZhukovskij) [reply client/src/services/translationService.js]: Done

File: client/src/services/translationService.js (hunk 25 above)
Comment (KonstantinZhukovskij) [reply client/src/services/translationService.js]: Done

**Pull request merged**